types/wlr_virtual_pointer_v1.c

```diff
@@ -0,0 +1,327 @@
+	if (pointer != NULL) {
+		wlr_input_device_destroy(&pointer->input_device);
+	}
+}
+
+static void virtual_pointer_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static const struct zwlr_virtual_pointer_v1_interface virtual_pointer_impl = {
+	.motion = virtual_pointer_motion,
+	.motion_absolute = virtual_pointer_motion_absolute,
+	.button = virtual_pointer_button,
+	.axis = virtual_pointer_axis,
+	.frame = virtual_pointer_frame,
+	.axis_source = virtual_pointer_axis_source,
+	.axis_stop = virtual_pointer_axis_stop,
+	.axis_discrete = virtual_pointer_axis_discrete,
+	.destroy = virtual_pointer_destroy,
+};
+
+static const struct zwlr_virtual_pointer_manager_v1_interface manager_impl;
+
+static struct wlr_virtual_pointer_manager_v1 *manager_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwlr_virtual_pointer_manager_v1_interface, &manager_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void virtual_pointer_manager_create_virtual_pointer(
+		struct wl_client *client, struct wl_resource *resource,
+		struct wl_resource *seat, uint32_t id) {
+	struct wlr_virtual_pointer_manager_v1 *manager = manager_from_resource(resource);
+
+	struct wlr_virtual_pointer_v1 *virtual_pointer = calloc(1,
+		sizeof(struct wlr_virtual_pointer_v1));
+	if (!virtual_pointer) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wlr_pointer *pointer = calloc(1, sizeof(struct wlr_pointer));
+	if (!pointer) {
+		wlr_log(WLR_ERROR, "Cannot allocate wlr_pointer");
+		free(virtual_pointer);
+		wl_client_post_no_memory(client);
+		return;
+	}
+	wlr_pointer_init(pointer, NULL);
+
+	struct wl_resource *pointer_resource = wl_resource_create(client,
+		&zwlr_virtual_pointer_v1_interface, wl_resource_get_version(resource),
+		id);
+	if (!pointer_resource) {
+		free(pointer);
+		free(virtual_pointer);
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_resource_set_implementation(pointer_resource, &virtual_pointer_impl,
+		virtual_pointer, virtual_pointer_destroy_resource);
+
+	wlr_input_device_init(&virtual_pointer->input_device,
+		WLR_INPUT_DEVICE_POINTER, &input_device_impl, "virtual pointer",
+		0x0, 0x0);
+
+	struct wlr_virtual_pointer_v1_new_pointer_event event = {
+		.new_pointer = virtual_pointer,
+	};
+
+	if (seat) {
+		struct wlr_seat_client *seat_client =
+			wlr_seat_client_from_resource(seat);
+		event.suggested_seat = seat_client->seat;
+	}
+
+	virtual_pointer->input_device.pointer = pointer;
+	virtual_pointer->resource = pointer_resource;
+	wl_signal_init(&virtual_pointer->events.destroy);
+
+	wl_list_insert(&manager->virtual_pointers, &virtual_pointer->link);
+	wlr_signal_emit_safe(&manager->events.new_virtual_pointer, &event);
+}
+
+static const struct zwlr_virtual_pointer_manager_v1_interface manager_impl = {
+	.create_virtual_pointer = virtual_pointer_manager_create_virtual_pointer
```

Comment:
We're missing a destroy request handler here.

Reply:
There is no destroy request.

Reply:
Hmm, we should definitely add one.